File: C/Algorithms/1624.c
#include "uthash/src/uthash.h"
#include <stdlib.h>

typedef struct {
  int key;
  int min;
  int max;
  UT_hash_handle hh;
} d_entry;

int maxLengthBetweenEqualCharacters(char *s) {
  // Build distances.
  d_entry *d = NULL;
  int i = 0;
  while (*s) {
    int chr = *s++;
    d_entry *t;
    HASH_FIND_INT(d, &chr, t);
    if (!t) {
      t = malloc(sizeof(*t));
      t->key = chr;
      t->min = i;
      t->max = i;
      HASH_ADD_INT(d, key, t);
    } else {
      if (i > t->max) {
        t->max = i;
      }
    }
    i++;
  }

  // Find max distance:
  d_entry *cur, *tmp;
  int max_dst = -1;
  HASH_ITER(hh, d, cur, tmp) {
    if (cur->min != cur->max) {
      int diff = cur->max - cur->min - 1;
      if (diff > max_dst)
        max_dst = diff;
    }
    HASH_DEL(d, cur);
  }
  return max_dst;
}

File: Javascript/Algorithms/1624.js
let maxLengthBetweenEqualCharacters = s => {
    const d = new Map,
          length = s.length;
    // Build distances.
    for(let i = 0; i < length; i++){
        let c = s[i];
        if (!d.has(c)){
            d.set(c, [i, i]);
        } else {
            let lst = d.get(c);
            if (i > lst[1]){
                lst[1] = i;
                d.set(c, lst);
            }
        }
    }
    // and then find max.
    let max_diff = -1;
    for ([min, max] of d.values()){
        if (min != max){
            let diff = max - min - 1;
            if (diff > max_diff){
                max_diff = diff;
            }
        }
    }
    return max_diff;
};
